### DISCUSSION TITLE
Support for X509 name constraints

### DISCUSSION THREAD

**Discussion by g-psantos** [strongswan/strongswan#1951] [Ideas]
## Proposed feature

Add support for the X509 `nameConstraints` extension.

According to this old [Redmine issue](https://wiki.strongswan.org/issues/3304) and [source code](https://github.com/strongswan/strongswan/blob/b5760243870d4ab6c5874d765f8b09b3c67b8fc6/src/libstrongswan/plugins/x509/x509_cert.c), it appears that support is already present in the X509 plugin but has to be added to the OpenSSL plugin.

## Background

strongSwan currently does not support technically constrained Issuing CAs.

A technically constrained CA is one that has the `nameConstraints` extension added to its certificate, with one or more `permitted` or `excluded` entries. For example, an Issuing CA can be technically constrained so that it may only issue certificates for `domain.com`, and clients will not trust any certificates issued by that CA for any other domain. X509 Name Constraints are a helpful tool in mitigating the potential for damage by a compromised CA.

If the `nameConstraints` extension on the certificate is marked "critical" (which it _should_ be, according to the CA/Browser Forum's Baseline Requirements), strongSwan discards the certificate when starting up:

```
found unsupported critical X.509 extension: X509v3 Name Constraints
OpenSSL X.509 parsing failed
building CRED_CERTIFICATE - X509 failed, tried 4 builders
```

The Issuing CA's certificate is thus not included in the certificate chain sent by strongSwan when a connection is configured with `pubkey` local authentication, and clients will fail to verify strongSwan's end-entity certificate unless they have the Intermediate CA certificate stored already. (This is unusual, as clients will typically only be configured to trust Root CAs.)

## Workarounds

I can work around the limitation by configuring the X509 extension to _not_ discard certificates for which it cannot parse critical extensions:

```
# ${sysconfdir}/strongswan.d/x509-critical.conf

libstrongswan {
  x509 {
    enforce_critical = no
  }
}
```

Of course, this is less-than-ideal because it'll defeat the purpose of having extensions marked "critical" in the first place.

**Comment by tobiasbrunner** [comment]
As you noted, you can always use the _x509_ plugin to parse certificates if you need support for special constraints. But adding support for the `nameConstraints` extension in the _openssl_ plugin is pretty straight forward (see the _1951-openssl-constraints_ branch). The other constraints that are supported by the _constraints_ plugin (`certificatePolicies`, `policyMappings`, `policyConstraints`, and `inhibitAnyPolicy`) might be trickier.

**Reply by g-psantos** [reply]
Thank you for the quick response and the code branch. Before I jump into custom builds that incorporate those changes:

- How can I use the _x509_ plugin in place of _openssl_? I tried increasing the priority of _x509_ (no change) and disabling _openssl_ altogether (certificates don't get parsed at all).

- Might the changes you put together be released with a future version? If so, I'll just keep the `enforce_critical` workaround for now and wait.

**Reply by tobiasbrunner** [reply]
> * How can I use the _x509_ plugin in place of _openssl_? I tried increasing the priority of _x509_ (no change) and disabling _openssl_ altogether (certificates don't get parsed at all).

Disabling the _openssl_ plugin might not work depending on the algorithms required for the certificate (or it will require several additional plugins to provide them). Changing the priority should work. However, you'll have to prioritize the _pem_ as well if the certificate is PEM-encoded. You can increase the log level for _asn_ to see details about the certificate parsing.


> * Might the changes you put together be released with a future version?

Sure, but if you could test them with your PKI that would be appreciated.

**Reply by g-psantos** [reply]
Thank you for your guidance and patience. I'll try to build that branch and test the changes using our PKI later this week and report back here.

As for the priority workaround, I tried to increase the _pem_ plugin priority as well and increasing the log level. It's still not working, but if I'm reading the log right, it seems that the cause is a set of IP address ranges for which the CA is allowed to issue certificates (basically, RFC1918 networks).

```log
# /var/log/strongswan.log
15[ASN2] L4 - extension:
15[ASN2] L5 - extnID:
15[ASN2]   'nameConstraints'
15[ASN2] L5 - critical:
15[ASN2]   TRUE
15[ASN2] L5 - extnValue:
15[ASN2] L6 - nameConstraints:
15[ASN2] L7 - permittedSubtrees:
15[ASN2] L8 - generalSubtree:
15[ASN2] L9 - dnsName:
15[ASN2]   'domain.com'
15[ASN2] L8 - generalSubtree:
15[ASN2] L9 - dnsName:
15[ASN2]   '.domain.com'
15[ASN2] L8 - generalSubtree:
15[ASN2] L9 - dnsName:
15[ASN2]   'domain.org'
15[ASN2] L8 - generalSubtree:
15[ASN2] L9 - dnsName:
15[ASN2]   '.domain.org'
15[ASN2] L8 - generalSubtree:
15[ASN2] L9 - ipAddress:
15[LIB2] builder L1 CRED_CERTIFICATE - X509 of plugin 'openssl'
15[LIB2] builder L2 CRED_PUBLIC_KEY - RSA of plugin 'pkcs1'
15[ASN2] L0 - RSAPublicKey:
15[ASN2] L1 - modulus:
15[ASN2] L1 - publicExponent:
15[LIB2] builder L3 CRED_PUBLIC_KEY - RSA of plugin 'pkcs1'
15[LIB2] builder L3 CRED_PUBLIC_KEY - RSA of plugin 'pgp'
15[LIB2] builder L3 CRED_PUBLIC_KEY - RSA of plugin 'dnskey'
15[LIB2] builder L3 CRED_PUBLIC_KEY - RSA of plugin 'openssl'
15[ASN2] L0 - algorithmIdentifier:
15[ASN2] L1 - algorithm:
15[ASN2]   'sha256WithRSAEncryption'
15[ASN2] L1 - parameters:
15[LIB1] found unsupported critical X.509 extension: X509v3 Name Constraints
15[LIB1] OpenSSL X.509 parsing failed
15[LIB2] builder L1 CRED_CERTIFICATE - X509 of plugin 'pem'
15[LIB1] building CRED_CERTIFICATE - X509 failed, tried 4 builders
```

```conf
# strongswan.conf
charon {
  plugins {
    x509 {
      load = 5
    }

    pem {
      load = 5
    }
  }

  filelog {
    asn {
      path = "/var/log/strongswan-asn.log"
      log_level = "yes"
      default = 2
    }
  }
}
```

**Reply by tobiasbrunner** [reply]
> I'm reading the log right, it seems that the cause is a set of IP address ranges for which the CA is allowed to issue certificates (basically, RFC1918 networks).

Yes, looks like parsing that constraint of type `iPAddress` fails, and as far as I can tell, this use case is currently not supported. The `generalName` parser in the _x509_ plugin only supports the basic form for this type (i.e. 4 or 16 bytes), not the extended form defined in RFC 5280 for `nameConstraints`, which refers to "address range" but actually just doubles the size by adding a netmask to denote a subnet. So unlike the format defined in RFC 3779, this only allows using proper subnets, no ranges, and wastes space in doing so. But we actually do have identification types that can be used for this, so I've pushed a bunch of commits to the branch that add support for this in the _x509_, _openssl_ and _constraints_ plugins.

**Reply by g-psantos** [reply]
Awesome, thank you so much. It'll probably take me a couple of days to get a custom build going and test the changes but I'll report back here.

**Reply by g-psantos** [reply]
Looks like that did the trick! I don't have a way to fully confirm with the charon/swanctl service, but at least _pki_ can parse the certificate:

```shell
$ cat /etc/swanctl/x509ca/cert.crt | pki --print          
  subject:  "[Redacted]"
  issuer:   "[Redacted]"
  validity:  not before [Redacted], ok
             not after  [Redacted], ok (expires in 1804 days)
  serial:    [Redacted]
  flags:     CA CRLSign serverAuth clientAuth ikeIntermediate 
  CRL URIs:  [Redacted]
  pathlen:   0
  permitted nameConstraints:
           domain.com
           .domain.com
           domain.org
           .domain.org
           10.0.0.0/8
           172.16.0.0/16
           192.168.0.0/16
  certificatePolicies:
             2.5.29.32.0
  authkeyId: [Redacted]
  subjkeyId: [Redacted]
  pubkey:    RSA 2048 bits
  keyid:     [Redacted]
  subjkey:   [Redacted]
```

<details>
<summary>My original idiocy (building from trunk and complaining it didn't work)</summary

**EDIT: Please hold on this -- I think I built strongSwan from the trunk after resetting the container (must've forgotten). Sorry!**

I was able to put together a custom build of that branch and imported the CA certificate into `/etc/swanctl/x509ca`.

I built this in a Docker container, so I can't launch strongSwan the normal way through systemd, and `pki --print` doesn't seem work (because _pki_ doesn't have the same plugins as the "main thing"):

```shell
$ cat /etc/swanctl/x509ca/cert.crt | pki --print
found unsupported critical X.509 extension: X509v3 Name Constraints
OpenSSL X.509 parsing failed
building CRED_CERTIFICATE - X509 failed, tried 4 builders
parsing input failed

$ pki --help
strongSwan 5.9.12dr1 PKI tool
loaded plugins: aes des rc2 sha2 sha1 md5 random x509 revocation pubkey pkcs1 pkcs7 pkcs12 dnskey sshkey pem openssl pkcs8 curve25519 hmac kdf drbg

$ swanctl --help
strongSwan 5.9.12dr1 swanctl
loaded plugins: aes des rc2 sha2 sha1 md5 random nonce x509 revocation constraints pubkey pkcs1 pkcs7 pkcs12 pgp dnskey sshkey pem openssl pkcs8 fips-prf curve25519 xcbc cmac hmac kdf gcm drbg
```

Is there an easy way to test just the part about loading up CA certificates?

</details>

**Reply by tobiasbrunner** [reply]
Great, thanks for this test. By the way, you can pass the plugins to load by `pki` via `PKI_PLUGINS` environment variable (space separated list of plugins). So you could basically use the list printed by the daemon to mirror its configuration and then use [`pki --verify`](https://docs.strongswan.org/docs/5.9/pki/pkiVerify.html) to check certificates.

**Reply by g-psantos** [reply]
To have your commits merged in, should I open an Issue referencing this Discussion or would you take it straight to a PR?

**Reply by tobiasbrunner** [reply]
No, you don't have to do anything.
